### PR TITLE
feat(DTFS2-6965): setup change stream for audit API

### PR DESCRIPTION
**Pull request by ChristianMcc**
## Introduction :pencil2:
We need to set up the change stream for mongodb to send updates to the audit API

## Resolution
Add change stream with a local mongodb replica set enabled. Add env variables to control if changestream is enabled and audit API variables.
When implementing the data changes we should think about if we want to explicitly exclude any collections e.g. the cron job collection from sending updates. I've set it up to automatically look for all collections so that any new collections that are added will be sent without extra work.

Note: I'll add a couple of tests for the changestream API call as there's some functionality in there

### CODE REVIEW

File: .github/workflows/test.yml
Comment (ChristianMcc) [inline .github/workflows/test.yml:128]: To be updated once env vars have been added to github secrets

File: utils/mock-data-loader/database/user-repository.js
Comment (AlexBramhill) [inline utils/mock-data-loader/database/user-repository.js:30]:
```suggestion
    const userCollection = await db.getCollection('users');
```

File: utils/mock-data-loader/database/user-repository.js
Comment (AlexBramhill) [inline utils/mock-data-loader/database/user-repository.js:38]: We're using `%o` elsewhere after Abhi's ticket is merged
```suggestion
    console.error('Failed to create %o', e);
```
fwiw as I had to look it up again:

If specifier is %o, optionally let converted be current with optimally useful formatting applied.

If specifier is %O, optionally let converted be current with generic JavaScript object formatting applied.

File: utils/mock-data-loader/database/user-repository.js
Comment (AlexBramhill) [inline utils/mock-data-loader/database/user-repository.js:39]: Will this just log `[Object.object]`? I'd like to start using `cause` for passing through the cause of the error ie


```suggestion
    throw new Error(`Failed to create logged in user session for user: ${user.username}`, {cause: e});
```

File: mongodb/Dockerfile
Comment (AlexBramhill) [inline mongodb/Dockerfile:4]: "Keyfiles are bare-minimum forms of security and are best suited for testing or development environments. For production environments we recommend using x.509 certificates." - https://www.mongodb.com/docs/manual/tutorial/enforce-keyfile-access-control-in-existing-replica-set/

Is `docker-compose.yml` our prod server?

Comment (AlexBramhill) [reply mongodb/Dockerfile:4]: Also - Is this creating the mongo keyfile on the docker image itself -- is this safe in terms of secret management?

File: dtfs-central-api/src/services/changeStream/changeStreamApi.ts
Comment (AlexBramhill) [inline dtfs-central-api/src/services/changeStream/changeStreamApi.ts:14]: We could probably just throw a new `EnvironmentalVariablesMissingError` here and then let the error handling in `setupChangeStream` handle it -- what do you think?

File: dtfs-central-api/src/services/changeStream/changeStreamApi.ts
Comment (AlexBramhill) [inline dtfs-central-api/src/services/changeStream/changeStreamApi.ts:43]: Are we going to be double logging this due to the same bit being in `setupChangeStream`?

File: dtfs-central-api/src/services/changeStream/changeStreamApi.test.ts
Comment (AlexBramhill) [inline dtfs-central-api/src/services/changeStream/changeStreamApi.test.ts:80]: Would this work for brevity?
```suggestion
    const envVarTestCases = [
      {envVarName: 'AUDIT_API_URL'},
      {envVarName: 'AUDIT_API_USERNAME'},
      {envVarName: 'AUDIT_API_PASSWORD'},
    ]
    it.each(envVarTestCases)('should not call the audit api if the %envVarName env var is missing', async ({envVarName}) => {
      delete process.env[envVarName];

      await postAuditDetails(mockChangeStreamDocument);

      expect(axios).not.toHaveBeenCalled();
    })
```

**Review by AlexBramhill — COMMENTED** [review]
lgtm after comments

File: utils/mock-data-loader/database/user-repository.js
Comment (ChristianMcc) [reply utils/mock-data-loader/database/user-repository.js:39]: This was some logging I didn't actually mean to commit. FWIW no it logged the full error but I'll delete!

File: utils/mock-data-loader/database/user-repository.js
Comment (ChristianMcc) [reply utils/mock-data-loader/database/user-repository.js:30]: lovely suggestion thank you, oops

File: mongodb/Dockerfile
Comment (ChristianMcc) [reply mongodb/Dockerfile:4]: This is purely for local dev yes - it's the keyfile that is used for the different mongodb instances in a replica set to talk to each other. When deployed azure takes care of all the replica set/keyfile issues for us and we don't need to worry about it :) It is safe as it's essentially throwaway in our local environment

File: dtfs-central-api/src/services/changeStream/changeStreamApi.test.ts
Comment (ChristianMcc) [reply dtfs-central-api/src/services/changeStream/changeStreamApi.test.ts:80]: Lovely suggestion :D

File: utils/mock-data-loader/database/user-repository.js
Comment (AlexBramhill) [reply utils/mock-data-loader/database/user-repository.js:39]: How odd , behaves differently in console in browser and on my local machine on an empty project 
image
! I'll file that along with how deep cloning works on this proj.

File: dtfs-central-api/src/services/changeStream/changeStreamApi.ts
Comment (ChristianMcc) [reply dtfs-central-api/src/services/changeStream/changeStreamApi.ts:14]: Didn't know this error existed! Will update :)

File: .github/workflows/test.yml
Comment (AlexBramhill) [reply .github/workflows/test.yml]: bump this comment

**Review by AlexBramhill — APPROVED** [review]
lgtm -- awaiting `CHANGE_STREAM_ENABLED` secret to be added to github secrets

File: .github/workflows/test.yml
Comment (AlexBramhill) [inline .github/workflows/test.yml:207]: Per offline conversation -- otherwise feel free to resolve and lgtm

**Review by AlexBramhill — APPROVED** [review]

**Review by AlexBramhill — APPROVED** [review]
Per offline conversation -- one comment I'm happy for you to resolve if I'm incorrect. Delightful work!

File: .github/workflows/test.yml
Comment (ChristianMcc) [reply .github/workflows/test.yml]: Have deleted :) was from when I was testing something and didn't undo - good spot

**Review by AlexBramhill — APPROVED** [review]
Reminder there is still a "To be updated once env vars have been added to github secrets" comment in `test.yml`

File: .github/workflows/test.yml
Comment (ChristianMcc) [reply .github/workflows/test.yml]: outdated :)

File: mongodb/Dockerfile
Comment (abhi-markan) [inline mongodb/Dockerfile:8]: Just out of curiosity, I suppose `999` is Mongo user?

File: utils/mock-data-loader/database/user-repository.js
Comment (abhi-markan) [inline utils/mock-data-loader/database/user-repository.js:34]: Can change to `Failed to create logged in user session %o` please.

File: dtfs-central-api/src/services/changeStream/changeStreamApi.ts
Comment (ChristianMcc) [reply dtfs-central-api/src/services/changeStream/changeStreamApi.ts:28]: Yeah could be a helper in future - I don't know if there's many places we use basic authentication with an API anywhere but can commonise if needed :)

File: dtfs-central-api/src/services/changeStream/changeStreamApi.ts
Comment (ChristianMcc) [reply dtfs-central-api/src/services/changeStream/changeStreamApi.ts:11]: Intentionally not wrapped in try-catch so that the calling function can handle any errors thrown :) The Promise<void> should be right, it doesn't return anything it's just async

File: dtfs-central-api/src/services/changeStream/setupChangeStream.ts
Comment (ChristianMcc) [reply dtfs-central-api/src/services/changeStream/setupChangeStream.ts:7]: Yeah this is a pattern where we aren't overriding the typescript type, and the typescript type shows it has no return value in this function

File: dtfs-central-api/src/services/changeStream/setupChangeStream.ts
Comment (ChristianMcc) [reply dtfs-central-api/src/services/changeStream/setupChangeStream.ts:15]: Error logging is all I'm doing for this right now, wouldn't want to crash the whole server if it fails but open to suggestions here

**Review by AlexBramhill — DISMISSED** [review]

File: dtfs-central-api/src/services/changeStream/setupChangeStream.ts
Comment (abhi-markan) [reply dtfs-central-api/src/services/changeStream/setupChangeStream.ts:15]: Makes sense for now. Thanks for the clarification.

**Review by abhi-markan — APPROVED** [review]

**Review by AlexBramhill — APPROVED** [review]

**Pull request merged**
